File: src/books/filters.py
# Katip Celebi
# Copyright (C) 2026 farukylmz0550
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Narrowing the library down, and putting it in order.

Deliberately free of Qt: what a filter *means* is worked out here and tested
here, and the widgets in grid.py only collect the answers.
"""

from dataclasses import dataclass

from books import tags
from books.model import Book, parse_rating, publish_year
from books.reading import STATUS_ANY, status_of
from shared.texts import text

SIGNED_ANY = "any"
SIGNED_YES = "yes"
SIGNED_NO = "no"

LENT_ANY = "any"
LENT_HOME = "home"
LENT_OUT = "out"

SORT_TITLE = "title"
SORT_RATING = "rating"
SORT_YEAR = "year"

# What the search box looks in. "All" is the everyday one; the rest are for
# when a word means different things in different places -- an author called
# Penguin, a book called Penguin, and Penguin the publisher.
SEARCH_ALL = "all"
SEARCH_TITLE = "title"
SEARCH_AUTHORS = "authors"
SEARCH_ISBN = "isbn"
SEARCH_PUBLISHERS = "publishers"
SEARCH_FIELDS = (
    SEARCH_ALL,
    SEARCH_TITLE,
    SEARCH_AUTHORS,
    SEARCH_ISBN,
    SEARCH_PUBLISHERS,
)

# What a ticked "signed" box writes. Stored as text, like every other field.
SIGNED_VALUE = "yes"
# Read generously: the library is a JSON file people do open and edit, and
# somebody typing "true" there plainly means the same as our "yes".
_MEANS_YES = {"yes", "true", "1", "evet"}


def is_signed(book: Book) -> bool:
    return book.signed.strip().lower() in _MEANS_YES


def haystack(book: Book, field: str, lent_out: bool = False) -> str:
    """The text one search looks through, folded for comparing.

    Searching everything is not the same as searching each field and joining
    the answers: "signed" and "lent out" are things you can type, and they are
    not written in the book anywhere.
    """
    if field == SEARCH_TITLE:
        parts = [book.title]
    elif field == SEARCH_AUTHORS:
        parts = [book.authors]
    elif field == SEARCH_ISBN:
        parts = [book.display_isbn, book.isbn_10, book.isbn_13]
    elif field == SEARCH_PUBLISHERS:
        parts = [book.publishers]
    else:
        parts = [
            book.authors,
            book.title,
            book.display_isbn,
            book.publishers,
            book.series,
            # As they read, not as they are stored: somebody searching for
            # "Science fiction" is typing what the card shows them.
            tags.show(book.tags),
            text("status_" + status_of(book)),
        ]
        if is_signed(book):
            # So typing "signed" finds them. One word only -- see the note on
            # signed_search_word in texts.py.
            parts.append(text("signed_search_word"))
        if lent_out:
            parts.append(text("lent_search_word"))
    return " ".join(part.casefold() for part in parts if part)


@dataclass(frozen=True)
class Filters:
    """What the user has narrowed the library down to."""

    query: str = ""
    search_field: str = SEARCH_ALL
    min_rating: int = 0
    signed: str = SIGNED_ANY
    lent: str = LENT_ANY
    status: str = STATUS_ANY
    tag: str = ""

    def allows(self, book: Book, lent_out: bool = False) -> bool:
        """Whether this book survives.

        ``lent_out`` is whether any copy is with somebody -- the ledger knows
        that and the book does not, so it is handed in.
        """
        wanted = self.query.strip().casefold()
        if wanted and wanted not in haystack(
            book, self.search_field, lent_out
        ):
            return False
        if self.min_rating and parse_rating(book.rating) < self.min_rating:
            return False
        if self.signed == SIGNED_YES and not is_signed(book):
            return False
        if self.signed == SIGNED_NO and is_signed(book):
            return False
        # A book you own three of, one of them out, is a book that is lent out.
        # Owning a spare does not make it not-lent: somebody still has one.
        if self.lent == LENT_OUT and not lent_out:
            return False
        if self.lent == LENT_HOME and lent_out:
            return False
        if self.status != STATUS_ANY and status_of(book) != self.status:
            return False
        if self.tag and not tags.contains(book.tags, self.tag):
            return False
        return True


def sort_key(book: Book, mode: str):
    """How to order one book.

    Every key is a tuple ending in the title, so books that tie -- and with
    ratings out of five, most of them tie -- come out alphabetically rather
    than in whatever order the file happened to hold them.
    """
    title = book.title.casefold()
    if mode == SORT_RATING:
        return (-parse_rating(book.rating), title)
    if mode == SORT_YEAR:
        return (-publish_year(book.publish_date), title)
    return (title,)


def arrange(
    books: list[Book], mode: str, descending: bool = False
) -> list[Book]:
    """The books in the order to show them."""
    ordered = sorted(books, key=lambda b: sort_key(b, mode))
    return list(reversed(ordered)) if descending else ordered

File: src/lib/books/filters.ts
// Filters — legacy src/books/filters.py port.
// Single responsibility: filter & sort.

export const SIGNED_ANY = "any";
export const SIGNED_YES = "yes";
export const SIGNED_NO = "no";
export const LENT_ANY = "any";
export const LENT_HOME = "home";
export const LENT_OUT = "out";
export const SORT_TITLE = "title";
export const SORT_RATING = "rating";
export const SORT_YEAR = "year";
export const SEARCH_ALL = "all";

export type Filters = {
  search: string;
  searchField: string; // all | title | authors | isbn | publishers
  minRating: number;
  signed: string; // any | yes | no
  lent: string; // any | home | out
  status: string; // any | not_read | want_to_read | reading | read
  tag: string;
  sort: string; // title | rating | year
  asc: boolean;
};

export const defaultFilters: Filters = {
  search: "",
  searchField: "all",
  minRating: 0,
  signed: SIGNED_ANY,
  lent: LENT_ANY,
  status: "any",
  tag: "any",
  sort: SORT_TITLE,
  asc: true,
};

export function isSigned(signed: boolean, filter: string): boolean {
  if (filter === SIGNED_ANY) return true;
  if (filter === SIGNED_YES) return signed;
  if (filter === SIGNED_NO) return !signed;
  return true;
}

export function haystack(book: { title: string; author?: string | null; isbn?: string | null; publishers?: string | null; tags?: string | null; signed?: boolean; status?: string }, lentOut: boolean): string {
  const parts = [
    book.title ?? "",
    book.author ?? "",
    book.isbn ?? "",
    book.publishers ?? "",
    book.tags ?? "",
    book.signed ? "yes" : "no",
    lentOut ? "out" : "home",
    book.status ?? "",
  ];
  return parts.join(" ").toLowerCase();
}

export function allows(book: { rating?: number | null; signed?: boolean; status?: string; tags?: string | null; title?: string; author?: string | null; isbn?: string | null; publishers?: string | null }, lentOut: boolean, f: Filters): boolean {
  if (f.minRating > 0 && (book.rating ?? 0) < f.minRating) return false;
  if (!isSigned(!!book.signed, f.signed)) return false;
  if (f.lent !== LENT_ANY) {
    if (f.lent === LENT_HOME && lentOut) return false;
    if (f.lent === LENT_OUT && !lentOut) return false;
  }
  if (f.status !== "any" && (book.status ?? "") !== f.status) return false;
  if (f.tag !== "any" && f.tag !== "" && !book.tags?.toLowerCase().includes(f.tag.toLowerCase())) return false;
  if (f.search.trim()) {
    const q = f.search.toLowerCase();
    const hay = haystack(book as never, lentOut);
    if (f.searchField === "title" && !book.title?.toLowerCase().includes(q)) return false;
    else if (f.searchField === "authors" && !book.author?.toLowerCase().includes(q)) return false;
    else if (f.searchField === "isbn" && !book.isbn?.toLowerCase().includes(q)) return false;
    else if (f.searchField === "publishers" && !book.publishers?.toLowerCase().includes(q)) return false;
    else if (f.searchField === SEARCH_ALL && !hay.includes(q)) return false;
  }
  return true;
}

export function sortKey(book: { title?: string; rating?: number | null; publishDate?: string | null }, sort: string): string | number {
  if (sort === SORT_RATING) return -(book.rating ?? 0);
  if (sort === SORT_YEAR) {
    const m = book.publishDate?.match(/(?<!\d)(\d{4})(?!\d)/);
    return m ? -parseInt(m[1], 10) : 0;
  }
  return (book.title ?? "").toLowerCase();
}

export function arrange<T extends { title?: string; id: string }>(books: T[], filters: Filters, lentMap: Map<string, boolean>): T[] {
  const filtered = books.filter((b) => allows(b as never, lentMap.get(b.id) ?? false, filters));
  const sorted = [...filtered].sort((a, b) => {
    const ka = sortKey(a as never, filters.sort);
    const kb = sortKey(b as never, filters.sort);
    if (ka < kb) return filters.asc ? -1 : 1;
    if (ka > kb) return filters.asc ? 1 : -1;
    // tie-break by title
    return (a.title ?? "").toLowerCase().localeCompare((b.title ?? "").toLowerCase());
  });
  return sorted;
}
